### DISCUSSION TITLE
How to log hyperparams, so they can be viewed in "HPARAMS" tab of Tensorboard?

### DISCUSSION THREAD

**Discussion by adosar** [Lightning-AI/pytorch-lightning#19960] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I am trying to log my hyperparameters with their metric values, so I can view them (compare runs with different hparams) under the "HPARAMS" tab of Tensorboard.

From the [docs](https://lightning.ai/docs/pytorch/stable/extensions/logging.html#logging-hyperparameters) (I don't know if there is something more relevant), I tried the following snippet:
```python
# Using custom or multiple metrics (default_hp_metric=False)
def on_train_start(self):
    self.logger.log_hyperparams(self.hparams, {"foo_metric": 0})

def validation_step(self, batch, batch_idx):
    self.log("foo_metric", some_scalar_1)
```

However, I am getting the following:
![Screenshot from 2024-06-08 22-45-10](https://github.com/Lightning-AI/pytorch-lightning/assets/110358278/bb711f50-d048-42cc-9ed2-6da8fc9b9f3f)

**TLDR**
Is there any way to get the output of [add_hparams](https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter.add_hparams) with Pytorch Lightning?

**Comment by mazurkin** [comment]
so check their code at `lightning_fabric.loggers.tensorboard.TensorBoardLogger.log_hyperparams`

for some reason they use `writer.add_summary` not `writer.add_hparams`, the comment says

> TensorBoard logs with and without saved hyperparameters are incompatible, the
> hyperparameters are then not displayed in the TensorBoard. Please delete or move the reviously saved > logs to display the new ones with hyperparameters.

I made my own subclass with proper implementation:

```python
class TensorBoardMultiLogger(pl.loggers.TensorBoardLogger):
    """TensorBoard logger that groups metrics into subdirectories by name prefix.

    Splits metric names on the first '/' separator: the prefix becomes a
    TensorBoard tag group and the suffix becomes the scalar name within that
    group. Metrics without a '/' are logged individually as-is.

    For example, 'loss/nt_xent/trn' is grouped under 'loss' with key
    'nt_xent/trn', so TensorBoard displays all loss variants together.
    """

    ROOT_SPLITTER: t.Final[str] = '/'

    @t.override
    @pl.utilities.rank_zero_only
    def log_metrics(self, metrics: t.Mapping[str, float], step: t.Optional[int] = None) -> None:
        assert pl.utilities.rank_zero_only.rank == 0, 'experiment tried to log from global_rank != 0'

        collection_metrics: dict[str, dict[str, float]] = collections.defaultdict(dict)
        individual_metrics: dict[str, float] = dict()

        for name, value in metrics.items():
            #  split the name into the metric root name and metric value name
            name_parts: list[str] = name.split(self.ROOT_SPLITTER, maxsplit=1)

            match len(name_parts):
                case 2:
                    collection_metrics[name_parts[0]][name_parts[1]] = value
                case 1:
                    individual_metrics[name] = value
                case name_parts_count:
                    raise RuntimeError('the name split failed (empty name?): ' + str(name_parts_count))

        # log all collection metrics
        # the method signature claim it does not support the nested dicts
        # but the implementation actually does (refer to the sources)
        # noinspection PyTypeChecker
        super().log_metrics(collection_metrics, step)

        # log individual metrics
        super().log_metrics(individual_metrics, step)

    @t.override
    @pl.utilities.rank_zero_only
    def log_hyperparams(
        self,
        params: dict[str, t.Any],
        metrics: t.Optional[dict[str, t.Any]] = None,
        step: t.Optional[int] = None,
    ) -> None:
        assert pl.utilities.rank_zero_only.rank == 0, 'experiment tried to log from global_rank != 0'

        # noinspection PyProtectedMember
        from lightning_fabric.utilities.logger import _add_prefix, _convert_params, _flatten_dict

        params = _convert_params(params)
        params = _flatten_dict(params)
        params = self._sanitize_params(params)

        metrics = _add_prefix(metrics, 'hparam', self.ROOT_SPLITTER)

        # noinspection PyTypeChecker
        return self.experiment.add_hparams(
            hparam_dict=params,
            metric_dict=metrics,
            global_step=step,
        )
```
